dvc/dependency/repo.py

```diff
@@ -112,64 +120,71 @@ def get_used_objs(
                 self.def_repo[self.PARAM_REV_LOCK] = rev
 
             path_info = PathInfo(repo.root_dir) / str(self.def_path)
-            try:
-                for odb, objs in repo.used_objs(
-                    [os.fspath(path_info)],
-                    force=True,
-                    jobs=kwargs.get("jobs"),
-                    recursive=True,
-                ).items():
-                    if odb is None:
-                        odb = repo.cloud.get_remote().odb
-                    self._check_circular_import(odb)
-                    used_objs[odb].update(objs)
-            except (NoRemoteError, NoOutputOrStageError):
-                pass
-
+            if not obj_only:
+                try:
+                    for odb, objs in repo.used_objs(
+                        [os.fspath(path_info)],
+                        force=True,
+                        jobs=kwargs.get("jobs"),
+                        recursive=True,
+                    ).items():
+                        if odb is None:
+                            odb = repo.cloud.get_remote().odb
+                            odb.read_only = True
+                        self._check_circular_import(odb, objs)
+                        used_objs[odb].update(objs)
+                except (NoRemoteError, NoOutputOrStageError):
+                    pass
+
+            staging = self.repo.odb.get_staging()
             staged_obj = stage(
-                local_odb,
+                staging,
                 path_info,
                 repo.repo_fs,
                 local_odb.fs.PARAM_CHECKSUM,
             )
+            staging.read_only = True
+
+            # NOTE: we cache unfiltered trees in _staged_objs but always
+            # return the filtered object in used_objs
             self._staged_objs[rev] = staged_obj
-            git_odb = GitObjectDB(repo.repo_fs, repo.root_dir)
-            used_objs[git_odb].add(staged_obj)
-            return used_objs
 
-    def _check_circular_import(self, odb):
+            filtered = self._filter_obj(staged_obj)
+            used_objs[staging].add(filtered)
+            return used_objs, filtered
+
+    def _check_circular_import(self, odb, objs):
```

Comment:
Is this checking for the possibility in all the chained imports?

Reply:
Yes, for chained imports the `erepo.used_objs()` call is recursive, so each "nested" erepo will also end up checking whether or not it has already been imported